test/CustomHashSetWithMemoryLimitTest.cpp

```diff
@@ -0,0 +1,229 @@
+      35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51,
+      52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68,
+      69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85,
+      86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100};
+  ASSERT_THROW(
+      {
+        for (auto num : testNums) {
+          hashSet.insert(num);
+        }
+      },
+      ad_utility::detail::AllocationExceedsLimitException);
+}
+
+TEST(CustomHashSetWithMemoryLimitTest, iteratorOperations) {
+  Set hashSet{makeAllocationMemoryLeftThreadsafeObject(1000_B)};
+  hashSet.insert(1);
+  hashSet.insert(2);
+  hashSet.insert(3);
+
+  // Test iterator functionality
+  auto it = hashSet.find(2);
+  ASSERT_NE(it, hashSet.end());
+  ASSERT_EQ(*it, 2);
+
+  // Test non-existing element
+  ASSERT_EQ(hashSet.find(4), hashSet.end());
+
+  // Test iteration
+  std::vector<int> values;
+  for (const auto& val : hashSet) {
+    values.push_back(val);
+  }
+  std::sort(values.begin(), values.end());
+  ASSERT_EQ(values, std::vector<int>({1, 2, 3}));
+}
+
+TEST(CustomHashSetWithMemoryLimitTest, eraseOperations) {
+  Set hashSet{makeAllocationMemoryLeftThreadsafeObject(1000_B)};
+  hashSet.insert(1);
+  hashSet.insert(2);
+  hashSet.insert(3);
+
+  // Test erase of existing element
+  hashSet.erase(2);
+  ASSERT_EQ(hashSet.size(), 2);
+  ASSERT_FALSE(hashSet.contains(2));
+
+  // Test erase of non-existing element
+  size_t originalSize = hashSet.size();
+  hashSet.erase(4);
+  ASSERT_EQ(hashSet.size(), originalSize);
+}
+
+TEST(CustomHashSetWithMemoryLimitTest, clearOperation) {
+  Set hashSet{makeAllocationMemoryLeftThreadsafeObject(1000_B)};
+  auto initialMemory = hashSet.getCurrentMemoryUsage();
+
+  // Add some elements
+  hashSet.insert(1);
+  hashSet.insert(2);
+  hashSet.insert(3);
+
+  auto usedMemory = hashSet.getCurrentMemoryUsage();
+  ASSERT_GT(usedMemory, initialMemory);
+
+  // Clear the set
+  hashSet.clear();
+  ASSERT_EQ(hashSet.size(), 0);
+  ASSERT_TRUE(hashSet.empty());
+
+  // Memory usage should be back to approximately initial state
+  // (might be slightly different due to bucket array size)
+  ASSERT_LE(hashSet.getCurrentMemoryUsage(), usedMemory);
+}
+
+TEST(CustomHashSetWithMemoryLimitTest, memoryTrackingAccuracy) {
+  Set hashSet{makeAllocationMemoryLeftThreadsafeObject(1000_B)};
+  auto initialMemory = hashSet.getCurrentMemoryUsage();
+
+  // Insert and track memory changes
+  hashSet.insert(1);
+  auto afterOneInsert = hashSet.getCurrentMemoryUsage();
+  ASSERT_GT(afterOneInsert, initialMemory);
+
+  // Insert duplicate and verify memory doesn't change
+  auto beforeDuplicate = hashSet.getCurrentMemoryUsage();
+  hashSet.insert(1);
+  ASSERT_EQ(hashSet.getCurrentMemoryUsage(), beforeDuplicate);
+
+  // Remove element and verify memory decreases
+  hashSet.erase(1);
+  ASSERT_EQ(hashSet.getCurrentMemoryUsage(), initialMemory);
+}
```

Comment:
The tracking for the test can also be performed by the proposed `MemoryCountingAllocator`.